packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -75,10 +75,26 @@ function preload(
       seen[dep] = true
       const isCss = dep.endsWith('.css')
       const cssSelector = isCss ? '[rel="stylesheet"]' : ''
-      // @ts-ignore check if the file is already preloaded by SSR markup
-      if (document.querySelector(`link[href="${dep}"]${cssSelector}`)) {
+      const isBaseRelative = !!importerUrl
+
+      // check if the file is already preloaded by SSR markup
+      if (isBaseRelative) {
+        const separatorIdx = dep.lastIndexOf('/')
+        const shortDep = separatorIdx < 0 ? dep : dep.slice(separatorIdx + 1)
+        const linkSelector = `link[href$="${shortDep}"]${cssSelector}`
+        const links = document.querySelectorAll<HTMLLinkElement>(linkSelector)
```

Comment:
For this particular optimization, we could use the `dep` before being passed through `assetsURL`. But once we merge https://github.com/vitejs/vite/pull/9938, `assetsURL` would no longer be always a relative path. 
I don't think we are saving much given the number of CSS link tags in an HTML, and we could be better off by getting every `link${cssSelector}`. What do you think?

Reply:
And thanks for the comment you added here, and your explanation at https://github.com/vitejs/vite/pull/9970#issuecomment-1244010551 🙏🏼

Reply:
> For this particular optimization, we could use the `dep` before being passed through `assetsURL`. But once we merge #9938, `assetsURL` would no longer be always a relative path. I don't think we are saving much given the number of CSS link tags in an HTML, and we could be better off by getting every `link${cssSelector}`. What do you think?

1. I didn't understand why #9938 conflicts with this method. This idea works no matter if you're in absolute path or in a relative one, it just compares the real file name without the "directory path", and it checks after if the absolute paths matches.
2. If you want this way, I suggest to query all the CSS links and all the JS links in the head of `preload`, so you won't need to query the same CSS matching request for every dep. (Or use memoization if you want to be fancy)

Reply:
Maybe the mistake @patak-dev and @bgoscinski you had, is that you thought the code was:
```js
const separatorIdx = dep.indexOf('/')
```
and it returns the first '/'. But no, it was `lastIndexOf('/')`, so it returns the *last* index of '/'. In the extreme, non-currently-happen edge case that '/' isn't present, as you can see it takes the full path, since no directory path is present.

Reply:
@Tal500 agree with you that the `dep.lastIndexOf('/')` will also work. What I say it won't is using `dep` before getting the full URL. But I still think that we won't win much by prefiltering the CSS links. And good idea on 2., we can do the query only once at the beginning of the function 👍🏼

Reply:
> @Tal500 agree with you that the `dep.lastIndexOf('/')` will also work. What I say it won't is using `dep` before getting the full URL.

Just didn't understand the new mechanism...
What do you mean? What will happen to the new `dep` path? Is the full path resolution will be so different that it breaks it? Sorry, just didn't understand the differences in the new way.

Reply:
Notice the first `$` in the `link[href$="${shortDep}"]` query(sorry if you already noticed). It means it selects all the links which their `href` value **ends with** `shotDep`.

I just can't see a situation this logic breaks, no matter what you're doing in the path, and what is the relative base dir.

Reply:
It won't break, the logic is always good. What I say won't work is using `dep` as it is before passing it through the `assetsURL` function WITHOUT doing the `dep.lastIndexOf('/')`. That know that I write it down, I see it won't work even before #9938. But don't focus on that, what I was trying to get at is that we aren't winning much for the complexity of this code. And as you proposed, it would be more efficient to do the query only once at the beginning of the function so it is better to query all the CSS links directly 👍🏼

Reply:
BTW: If we want to go with the "catch all the links at the start of preload" method, I suggest the following pre-processing, in the head of "preload":
1. Query all the links(JS&CSS).
2. Convert each link href into its **absolute path**
3. Create a JS `Set` object containing all the **absolute paths**

Then, for every dep, get its full path, and then check if it's already in the set.

This method is simple enough, and even might be even more optimized(as browser querying isn't free) than what I had suggested earlier.

Reply:
1. Why do you want to query JS links? Do we have a case where we see a duplicate? (even if there is, it isn't an issue like with CSS, no?)
2. Conversion would be just `links[i].href`, as @bgoscinski explained, no?
3. Having the set could be good 👍🏼, although I think the perf may end up being the same for the small list we will be dealing with here.

Reply:
@bgoscinski Do you agree to implement the pre-precessing&verifying, in the head of `preload, as I described in https://github.com/vitejs/vite/pull/9970#discussion_r968889115 ?

Reply:
> 1. Why do you want to query JS links? Do we have a case where we see a duplicate? (even if there is, it isn't an issue like with CSS, no?)
> 
>     2. Conversion would be just `links[i].href`, as @bgoscinski explained, no?
> 
>     3. Having the set could be good 👍🏼, although I think the perf may end up being the same for the small list we will be dealing with here.

1. Don't know, I thought everything is possible, many paths leads to the same place(you know Vite ways better though).
2. Yes
3. Not only it's efficient, it's also has a simple JS API to query if something where already seen.

If we where more crazy, I would have suggested Bloom-Filter🙃. (Only a joke, we're in JS and small data world)